packages/loader/container-utils/src/liveSummarySanitizer.ts

```diff
@@ -0,0 +1,194 @@
+        keys.includes(i),
+    );
+    return filtered.length === keys.length;
+}
+
+function createBlobItem(content: any): ISummaryBlob {
+    return {
+        content: JSON.stringify(content),
+        type: SummaryType.Blob,
+    };
+}
+
+function validateTree(liveSummary: ISummaryTree): boolean {
+    for (const key of Object.keys(liveSummary.tree)) {
+        const o = liveSummary.tree[key];
+        if (o.type === SummaryType.Handle) {
+            return false;
+        } else if (o.type === SummaryType.Tree && !validateTree(o)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+class LiveV1SummarySanitizer {
+    private readonly quorumKeyIdx = 0;
+    private readonly quorumValIdx = 1;
+    private readonly emptySummaryBlob: ISummaryBlob = {
+        content: JSON.stringify([]),
+        type: SummaryType.Blob,
+    };
+
+    constructor(private readonly srcSummary: ISummaryTree) {}
+
+    public get sanitizedCopy(): ISummaryTree {
+        const appTree = this.sanitizeAppTree(this.srcSummary);
+        const protocolTree = this.sanitizeProtocolTree(this.srcSummary);
+        return {
+            type: SummaryType.Tree,
+            tree: {
+                ".app": appTree,
+                ".protocol": protocolTree,
+            },
+        };
+    }
+
+    /**
+     * Sanitize app tree
+     * @param srcSummary - summary to transform
+     */
+    private sanitizeAppTree(srcSummary: ISummaryTree): ISummaryTree {
+        const keys = [".channels", ".electedSummarizer", ".metadata"];
+        assert(
+            validateKeyPresence(keys, srcSummary),
+            "Valid app tree keys should be present",
+        );
+
+        const parentTree = srcSummary.tree;
+        const meta = readBlobContent(
+            (parentTree[".metadata"] as ISummaryBlob).content,
+        );
+        assert(meta !== undefined, "Invalid app metadata");
+
+        return {
+            type: SummaryType.Tree,
+            tree: {
+                [".channels"]: parentTree[".channels"],
+                [".electedSummarizer"]: createBlobItem({
+                    electionSequenceNumber: 0,
+                }),
+                [".metadata"]: createBlobItem({
+                    createContainerRuntimeVersion:
+                        meta.createContainerRuntimeVersion,
+                    createContainerTimestamp: meta.createContainerTimestamp,
+                    summaryFormatVersion: meta.summaryFormatVersion,
+                }),
+            },
+        };
+    }
+
+    /**
+     * Sanitize protocol tree
+     * @param srcSummary - summary to transform
+     */
+    private sanitizeProtocolTree(srcSummary: ISummaryTree): ISummaryTree {
+        const protocolSummary = srcSummary.tree[".protocol"];
+        assert(
+            protocolSummary.type === SummaryType.Tree,
+            "Invalid object type",
+        );
+
+        const keys = [
+            "attributes",
```

Comment:
Curious why you need to even know the format of quorum and make any modifications? Not very clear from code (comments would help).
But if you do, I'd suggest keeping this code next to Quorum class itself and call it from here. Quorum snapshot format should not be leaked out and referenced across repo, this makes it really hard to make changes, it should be localized in one place

Reply:
I believe when we rehydrate from snapshot, we do not re-init back some of these fields. Similar comment as above. I'll double check this, and move it closer Quorum if we need them. Both container and sanitizer logic can point to the same utils. 

I originally kept it here thinking that we will be sanitizing (snapshot version) v1-to-v1, and in future v2-to-v2, without needing to spread that version-related logic elsewhere. Given we may need separate sanitization (and version handling) for DDS-es as well, it makes sense to start moving some of these transformations closer to models they are transforming. And basically always land transformation back to the newest snapshot version.

Reply:
V3 will need to be able to read and understand V1 & V2 in some form or shape (directly or through converters). 
https://microsoft-my.sharepoint-df.com/:w:/p/vladsud/Ea7aDhr2tRZHr65HYviWQP8B39Clfe0_PP41ULWYnpjyEg?e=JY0J8o